### PR TITLE
Add `set_status` for `std::process::Child`

### LINKED ISSUE #39188: Allow `process::Child` to be used with external calls to `waitpid`
The `wait` and `wait_with_output` methods of `process::Child` expect to do the waiting for the child to exit themselves.  If you are running many processes at once, and you don't know which one will finish first, this is inconvenient, because the system calls `wait` and `waitpid` consume each exit notification as they report it.  (`waitid` can be told not to do that, but it is not fully supported in the libc crate yet.)  That places the `Child` object in an inconsistent state, in which it is not safe to call `wait` or `wait_with_output`.

If `Child` had a `set_exit_status` method, an application in this situation could use it to keep the objects consistent:  something like

```rust
fn wait_all(children: &[&Child]) -> nix::Result<()> {
    use nix::sys::wait::{wait, WaitStatus}
    let mut pending : HashMap<u32, &Child> =
        children.iter().map(|c| (c.id(), c)).collect();
   while pending.len() {
       let status = try!(wait());
       match status {
           Exited(pid, ..) | Signaled(pid, ..) => {
               let child = try!(pending.remove(pid).ok_or(UnexpectedChild(pid)));
               child.set_exit_status(status); 
           },
           _ => unreachable!();
        }
    }
    Ok(())
}
```

(In this hypothetical, `ExitStatus` impls `From<nix::WaitStatus>`, which I believe nix can make happen without assistance from std.)

### PR DESCRIPTION
**Pull request by dylni**
Closes #39188

Are tests other than the doctest needed? Any good test would need to be very platform-specific.

I also removed `ExitStatus::new`, since it was only used for some platforms and could be replaced with constructing the struct directly.

### CODE REVIEW

**Comment by dylni** [comment]
Reassigning to another libs member. I don't think @withoutboats has been reviewing recently.

r? @sfackler

File: src/libstd/process.rs [hunk 1]
```diff
@@ -1489,6 +1489,46 @@ impl Child {
...
+    /// let status = external_wait(&mut child);
+    /// unsafe { child.set_status(status) }
+    ///
+    /// assert_eq!(status, child.wait().expect("failed to wait on child"));
+    /// ```
+    ///
+    /// [`wait`]: #method.wait
+    /// [`wait_with_output`]: #method.wait_with_output
+    #[stable(feature = "child_set_status", since = "1.44.0")]
+    pub unsafe fn set_status(&mut self, status: ExitStatus) {
```
Comment (sfackler) [inline src/libstd/process.rs]: In what way can calling this method incorrectly cause memory unsafety?

Comment (dylni) [reply src/libstd/process.rs]: My original reasoning was that [`kill` would be unexpectedly affected](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/process/process_unix.rs#L419-L421), and there could be other uses by `Child` in the future. However, thinking through it again, it does make sense to remove the `unsafe`.

**Comment by sfackler** [comment]
I'm somewhat worried that this opens too many boxes about the internal representation of Child - thoughts @rust-lang/libs?

**Comment by dylni** [comment]
If there's a warning that setting it to an inaccurate exit status results in undefined behavior, there shouldn't be a problem. It would require unsafe code to get the exit status externally anyway.

However, after updating the wait implementation in my crate, I actually no longer need to call this function. So, I am not entirely opposed to delaying it if there are objections.

**Comment by sfackler** [comment]
Ah cool yeah, if your code isn't going to be depending on the functionality, then I think it makes sense to close this for now then.

**Comment by dylni** [comment]
That works. Thanks for the review anyway!

**Pull request closed without merging**